gusto/time_discretisation.py

```diff
@@ -238,6 +239,202 @@ def apply(self, x_out, x_in):
+            butcher_exp (numpy array): A matrix containing the coefficients of
+                a butcher tableau defining a given explicit Runge Kutta time discretisation.
+            field_name (str, optional): name of the field to be evolved.
+                Defaults to None.
+            subcycles (int, optional): the number of sub-steps to perform.
+                Defaults to None.
+            solver_parameters (dict, optional): dictionary of parameters to
+                pass to the underlying solver. Defaults to None.
+            limiter (:class:`Limiter` object, optional): a limiter to apply to
+                the evolving field to enforce monotonicity. Defaults to None.
+            options (:class:`AdvectionOptions`, optional): an object containing
+                options to either be passed to the spatial discretisation, or
+                to control the "wrapper" methods, such as Embedded DG or a
+                recovery method. Defaults to None.
+        """
+        super().__init__(domain, field_name=field_name,
+                         solver_parameters=solver_parameters,
+                         limiter=limiter, options=options)
+        self.butcher_imp = butcher_imp
+        self.butcher_exp = butcher_exp
+        self.nStages = int(np.shape(self.butcher_imp)[1])
+
+    def setup(self, equation, apply_bcs=True, *active_labels):
+        """
+        Set up the time discretisation based on the equation.
+
+        Args:
+            equation (:class:`PrognosticEquation`): the model's equation.
+            *active_labels (:class:`Label`): labels indicating which terms of
+                the equation to include.
+        """
+
+        super().setup(equation, apply_bcs, *active_labels)
+
+        self.residual = self.residual.label_map(
+            lambda t: any(t.has_label(time_derivative, transport)),
+            map_if_false=lambda t: implicit(t))
+
+        self.residual = self.residual.label_map(
+            lambda t: t.has_label(transport),
+            map_if_true=lambda t: explicit(t))
+
+        self.xs = [Function(self.fs) for i in range(self.nStages)]
+
+    @cached_property
+    def lhs(self):
+        """Set up the discretisation's left hand side (the time derivative)."""
+        return super(IMEXMultistage, self).lhs
+
+    @cached_property
+    def rhs(self):
+        """Set up the discretisation's right hand side (the time derivative)."""
+        return super(IMEXMultistage, self).rhs
+
+    def res(self, stage):
+        mass_form = self.residual.label_map(
+            lambda t: t.has_label(time_derivative),
+            map_if_false=drop)
+        residual = mass_form.label_map(all_terms,
+                                       map_if_true=replace_subject(self.x_out, old_idx=self.idx))
+        residual -= mass_form.label_map(all_terms,
+                                        map_if_true=replace_subject(self.x1, old_idx=self.idx))
+        for i in range(stage):
```

Comment:
I can follow what is going on in this for loop but maybe you could add some comments to explain it?

Reply:
Good point! Comments added